### PR TITLE
Add `GF2Square` bloq for squaring over GF($2^m$)

### PR DESCRIPTION
**Pull request by tanujkhattar**
Part of a larger effort to support galois field arithmetic in qualtran. xref https://github.com/quantumlib/Qualtran/issues/1419

Currently on top of https://github.com/quantumlib/Qualtran/pull/1438

### CODE REVIEW

File: qualtran/bloqs/gf_arithmetic/gf2_square.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    @cached_property
+    def signature(self) -> 'Signature':
+        return Signature([Register('x', dtype=self.qgf)])
+
+    @cached_property
+    def qgf(self) -> QGF:
+        return QGF(characteristic=2, degree=self.bitsize)
+
+    @cached_property
+    def squaring_matrix(self) -> np.ndarray:
```
Comment (mpharrigan) [inline qualtran/bloqs/gf_arithmetic/gf2_square.py:68]: docstring

File: qualtran/bloqs/gf_arithmetic/gf2_square_test.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+def test_gf2_multiplication_symbolic(bloq_autotester):
+    bloq_autotester(_gf2_square_symbolic)
+
+
+def test_gf2_multiplication_classical_sim_quick():
+    m = 2
+    bloq = GF2Square(m)
+    GFM = GF(2**m)
+    assert_consistent_classical_action(bloq, x=GFM.elements)
+
```
Comment (mpharrigan) [inline qualtran/bloqs/gf_arithmetic/gf2_square_test.py:35]: optional test idea: assert cost is only cliffords

**Review by mpharrigan — APPROVED** [review]
lgtm

File: qualtran/bloqs/gf_arithmetic/gf2_square.py (hunk 1 above)
Comment (tanujkhattar) [reply qualtran/bloqs/gf_arithmetic/gf2_square.py:68]: done

File: qualtran/bloqs/gf_arithmetic/gf2_square_test.py (hunk 2 above)
Comment (tanujkhattar) [reply qualtran/bloqs/gf_arithmetic/gf2_square_test.py:36]: done

**Pull request merged**